plugins/http_plugin/tests/unit_tests.cpp

```diff
@@ -217,50 +223,98 @@ void run_test(Protocol& p, size_t max_body_size)
       test_str.resize(max_body_size + 1, '1');
       check_request(p, "/check_ones", test_str.c_str(), {}); // we don't expect a response
    }
-}   
+}
+
+namespace eosio {
+class chain_api_plugin : public appbase::plugin<chain_api_plugin> {
+ public:
+   APPBASE_PLUGIN_REQUIRES();
+   virtual void set_program_options(options_description& cli, options_description& cfg) override {}
+   void         plugin_initialize(const variables_map& options) {}
+   void         plugin_startup() {}
+   void         plugin_shutdown() {}
+};
+
+class net_api_plugin : public appbase::plugin<net_api_plugin> {
+ public:
+   APPBASE_PLUGIN_REQUIRES();
+   virtual void set_program_options(options_description& cli, options_description& cfg) override {}
+   void         plugin_initialize(const variables_map& options) {}
+   void         plugin_startup() {}
+   void         plugin_shutdown() {}
+};
+
+class producer_api_plugin : public appbase::plugin<producer_api_plugin> {
+ public:
+   APPBASE_PLUGIN_REQUIRES();
+   virtual void set_program_options(options_description& cli, options_description& cfg) override {}
+   void         plugin_initialize(const variables_map& options) {}
+   void         plugin_startup() {}
+   void         plugin_shutdown() {}
+};
+
+static auto _chain_api_plugin    = application::register_plugin<chain_api_plugin>();
+static auto _net_api_plugin      = application::register_plugin<net_api_plugin>();
+static auto _producer_api_plugin = application::register_plugin<producer_api_plugin>();
+} // namespace eosio
+
+struct http_plugin_test_fixture {
+   appbase::scoped_app app;
+   std::thread         app_thread;
+
+   http_plugin* init(std::initializer_list<const char*> args) {
+      if (app->initialize<http_plugin>(args.size(), const_cast<char**>(args.begin()))) {
+         auto                       plugin  = app->find_plugin<http_plugin>();
+         std::atomic<bool>& listening_or_failed = plugin->listening();
+         app_thread = std::thread([&]() {
+            try {
+               app->startup();
+               app->exec();
+            } catch (...) {
+               plugin = nullptr;
+               listening_or_failed.store(true);
+            }
+         });
+
+         while (!listening_or_failed.load()) {
+            using namespace std::chrono_literals;
+            std::this_thread::sleep_for(10ms);
+         }
+         return plugin;
```

Comment:
This should work. Maybe appbase infrastructure should hold a `std::promise` for each plugin, which would be provided a value after the plugin successfully completes its startup.